### ISSUE TITLE
proper c api header

### ISSUE THREAD

**Issue by CapsAdmin** [KhronosGroup/glslang#184]
at the moment the c api is included inline with c++ code (glslang/Public/ShaderLang.h) and contains some code that is invalid C such as default argument parameters and use of structs and enums as if they are typedefd.

**Comment by krux02** [comment]
yes this is an issue to me, too. Additionally, make install should also install those development headers, otherwise having the static libraries is pretty useless.

**Comment by dneto0** [comment]
You might want to look into https://github.com/google/shaderc
It's a C and C++ API wrapper around Glslang for compiling into SPIR-V.  The C++ part is headers-only implementation on top of the C API.
It also has glslc, a command line compiler that mimics GCC and Clang options and file handling, so it more naturally fits into traditional build systems.  It also implements an #include feature.

**Comment by Groovounet** [comment]
I would add: a proper C interface and a configuration to build dynamic library: Having to build static libraries per compiler version is not great.

**Comment by DagAgren** [comment]
Will this ever happen? It seems like it should be a very small change. Would a PR for it be accepted?

It's weird to have a C API that is not actually accessible from C.

**Comment by johnkslang** [comment]
A few years ago, for different reasons, we (people investing in glslang) decided to deprecate the C API in favor of a newer C++ API that has more functionality.

So, it would make sense, given the requests here, to know what all is desired to exposed through a C API, and do that (or check first if the wrapper @dneto0 mentioned does actually expose everything desired through its C API).

For example, do the needs include the reflection APIs?

**Comment by DagAgren** [comment]
Personally, all I want is to throw a string at it and get some SPIR-V to pass onwards.

**Comment by corporateshark** [comment]
Implemented here #2038

**Comment by johnkslang** [comment]
This is now merged. Let's try it as the new C API, and make incremental issues/fixes from there.

**Comment by enkore** [comment]
One thing that seems to be missing from the C API is passing a list of strings as the source code (like in glslang itself, or the legacy glShaderSource). I assume glslang is like the other glsl compilers in that embedded new-lines in a source string don't count as a new line for diagnostics?

If so, adding this seems sensible, if possible from a API stability point of view.

**Comment by corporateshark** [comment]
@enkore Why not to do something like this externally (in C) and use the resulting string? No need to bloat the API.

```
string src;
for (const auto& s : strings)
{
   src += s + '\n';
}
```

**Comment by johnkslang** [comment]
> ... that embedded new-lines in a source string don't count as a new line for diagnostics?

Internally, in glslang, you can certainly have newlines within a big string, so that a multi-line program is in one string, but still processed as multi-line, and you get line numbers on errors/warnings by counting the number of newlines seen so far in the string being processed.

(There are though subtleties around final nul characters.)

Internally, glslang should process a set of strings identically to as if any final nul characters were removed and the strings were concatenated, without introducing any extra newlines.

Maybe its a minor thing, but glslang can process the entire input without making any copies, and that might be nice to preserve.  Since the OpenGL API takes a set of strings, glslang also can process a set of strings, with no copying.

**Comment by arcady-lunarg** [comment]
This has been implemented in #2038, if there are further concerns or suggestions for improving the C API, feel free to open a separate issue for that.
